engine/src/main/java/org/terasology/engine/core/modes/AbstractState.java

```diff
@@ -0,0 +1,84 @@
+// Copyright 2021 The Terasology Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+package org.terasology.engine.core.modes;
+
+import org.terasology.engine.context.Context;
+import org.terasology.engine.core.ComponentSystemManager;
+import org.terasology.engine.core.bootstrap.EntitySystemSetupUtil;
+import org.terasology.engine.entitySystem.entity.EntityRef;
+import org.terasology.engine.entitySystem.entity.internal.EngineEntityManager;
+import org.terasology.engine.entitySystem.event.internal.EventSystem;
+import org.terasology.engine.logic.console.Console;
+import org.terasology.engine.logic.console.ConsoleImpl;
+import org.terasology.engine.logic.console.ConsoleSystem;
+import org.terasology.engine.logic.console.commands.CoreCommands;
+import org.terasology.engine.logic.players.LocalPlayer;
+import org.terasology.engine.network.ClientComponent;
+import org.terasology.engine.recording.DirectionAndOriginPosRecorderList;
+import org.terasology.engine.recording.RecordAndReplayCurrentStatus;
+import org.terasology.engine.registry.CoreRegistry;
+import org.terasology.engine.rendering.nui.NUIManager;
+import org.terasology.engine.rendering.nui.internal.NUIManagerInternal;
+import org.terasology.engine.rendering.nui.internal.TerasologyCanvasRenderer;
+import org.terasology.nui.canvas.CanvasRenderer;
+
+import static com.google.common.base.Verify.verifyNotNull;
+
+public abstract class AbstractState implements GameState {
+    protected Context context;
+    protected EngineEntityManager entityManager;
+    protected EventSystem eventSystem;
+    protected ComponentSystemManager componentSystemManager;
+
+    protected void initEntityAndComponentManagers() {
```

Comment:
This does a lot with the `context` and strictly requires it. Would it make sense to make the `context` and argument of this method, similar to how we pass it to `createLocalPlayer` explcitly?

Reply:
`createLocalPlayer` is the method I extracted first. Passing `context` let me make it `static`, which I figured would make it easier to move the method later if there were more copies of it.

Then I realized if I shuffled some things around I could also extract all this stuff that went in to `initEntityAndComponentManagers`, but because it assigns things to multiple instance fields, it wouldn't work as static.

Upon review, I see that all these fields are direct references to a thing in the context, so it probably could be split in to a "populate `context` with this stuff" static method and then a "pull some references out of context and cache them in fields" non-static method, but I worried that would make it more difficult to review if I included it in the same PR.